spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/MappingMongoJsonSchemaCreator.java

```diff
@@ -0,0 +1,213 @@
+		Class<?> rawTargetType = computeTargetType(property); // target type before conversion
+		Class<?> targetType = converter.computeWriteTarget(rawTargetType); // conversion target type
+
+		if (property.isEntity() && ObjectUtils.nullSafeEquals(rawTargetType, targetType)) {
+			return createObjectSchemaPropertyForEntity(path, property, required);
+		}
+
+		String fieldName = computePropertyFieldName(property);
+
+		if (property.isCollectionLike()) {
+			return createSchemaProperty(fieldName, targetType, required);
+		} else if (property.isMap()) {
+			return createSchemaProperty(fieldName, Type.objectType(), required);
+		} else if (ClassUtils.isAssignable(Enum.class, targetType)) {
+			return createEnumSchemaProperty(fieldName, targetType, required);
+		}
+
+		return createSchemaProperty(fieldName, targetType, required);
+	}
+
+	private JsonSchemaProperty createObjectSchemaPropertyForEntity(List<PersistentProperty> path,
+			PersistentProperty property, boolean required) {
+
+		ObjectJsonSchemaProperty target = JsonSchemaProperty.object(property.getName());
+		List<JsonSchemaProperty> nestedProperties = computePropertiesForEntity(path,
+				mappingContext.getPersistentEntity(property));
+
+		return createPotentiallyRequiredSchemaProperty(
+				target.properties(nestedProperties.toArray(new JsonSchemaProperty[0])), required);
+	}
+
+	private JsonSchemaProperty createEnumSchemaProperty(String fieldName, Class<?> targetType, boolean required) {
+
+		List<Object> possibleValues = new ArrayList<>();
+		for (Object enumValue : EnumSet.allOf((Class) targetType)) {
+			possibleValues.add(converter.convertToMongoType(enumValue));
+		}
+
+		targetType = possibleValues.isEmpty() ? targetType : possibleValues.iterator().next().getClass();
+		return createSchemaProperty(fieldName, targetType, required, possibleValues);
+	}
+
+	JsonSchemaProperty createSchemaProperty(String fieldName, Object type, boolean required) {
+		return createSchemaProperty(fieldName, type, required, Collections.emptyList());
+	}
+
+	JsonSchemaProperty createSchemaProperty(String fieldName, Object type, boolean required,
+			Collection<?> possibleValues) {
+
+		TypedJsonSchemaObject schemaObject = type instanceof Type ? JsonSchemaObject.of(Type.class.cast(type))
+				: JsonSchemaObject.of(Class.class.cast(type));
+
+		if (!CollectionUtils.isEmpty(possibleValues)) {
+			schemaObject = schemaObject.possibleValues(possibleValues);
+		}
+
+		return createPotentiallyRequiredSchemaProperty(JsonSchemaProperty.named(fieldName).with(schemaObject), required);
+	}
+
+	private String computePropertyFieldName(PersistentProperty property) {
+
+		return property instanceof MongoPersistentProperty ? ((MongoPersistentProperty) property).getFieldName()
+				: property.getName();
+	}
+
+	private boolean isRequiredProperty(PersistentEntity<?, ?> parent, PersistentProperty property) {
+
+		return (parent.isConstructorArgument(property) && !property.isAnnotationPresent(Nullable.class))
```

Comment:
See `NullableUtils`: Properties are nullable by default. Annotating a package with  `@NonNullApi` or `@NonNullFields` switches to non-null by default.